Тема №23 "Употребление алкоголя учащимися" - Балашов КМБО-03-20 

Набор данных№1- https://www.kaggle.com/uciml/student-alcohol-consumption?select=student-por.csv

Набор данных№2- https://www.kaggle.com/uciml/student-alcohol-consumption?select=student-mat.csv

Устанавливаем необходимые библиотеки:

In [1]:
!pip install pandas
!pip install sklearn
!pip install statsmodels 

In [2]:
import pandas 
import numpy 
import math
from sklearn.model_selection import train_test_split
from statsmodels.stats.outliers_influence import variance_inflation_factor
from sklearn import preprocessing
from sklearn.decomposition import PCA

Считываем наборы данных:

In [3]:
data1 = pandas.read_csv('student-por.csv')
data2 = pandas.read_csv('student-mat.csv')
Data = [data1, data1]
Data = pandas.concat(Data)

In [4]:
Data

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0       GP   F   18       U     GT3       A     4     4   at_home   teacher   
1       GP   F   17       U     GT3       T     1     1   at_home     other   
2       GP   F   15       U     LE3       T     1     1   at_home     other   
3       GP   F   15       U     GT3       T     4     2    health  services   
4       GP   F   16       U     GT3       T     3     3     other     other   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
644     MS   F   19       R     GT3       T     2     3  services     other   
645     MS   F   18       U     LE3       T     3     1   teacher  services   
646     MS   F   18       U     GT3       T     1     1     other     other   
647     MS   M   17       U     LE3       T     3     1  services  services   
648     MS   M   18       R     LE3       T     3     2  services     other   

     ... famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0    ...      4        3      4     1     1      3        4   0  11  11  
1    ...      5        3      3     1     1      3        2   9  11  11  
2    ...      4        3      2     2     3      3        6  12  13  12  
3    ...      3        2      2     1     1      5        0  14  14  14  
4    ...      4        3      2     1     2      5        0  11  13  13  
..   ...    ...      ...    ...   ...   ...    ...      ...  ..  ..  ..  
644  ...      5        4      2     1     2      5        4  10  11  10  
645  ...      4        3      4     1     1      1        4  15  15  16  
646  ...      1        1      1     1     1      5        6  11  12   9  
647  ...      2        4      5     3     4      2        6  10  10  10  
648  ...      4        4      1     3     4      5        4  10  11  11  

[1298 rows x 33 columns]

#Поскольку количество строк равно 1298, а последний индекс строки равен 648, то индексы второго датасета идут с нуля, а не с 396

In [5]:
data1.info()
data2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 649 entries, 0 to 648
Data columns (total 33 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   school      649 non-null    object
 1   sex         649 non-null    object
 2   age         649 non-null    int64 
 3   address     649 non-null    object
 4   famsize     649 non-null    object
 5   Pstatus     649 non-null    object
 6   Medu        649 non-null    int64 
 7   Fedu        649 non-null    int64 
 8   Mjob        649 non-null    object
 9   Fjob        649 non-null    object
 10  reason      649 non-null    object
 11  guardian    649 non-null    object
 12  traveltime  649 non-null    int64 
 13  studytime   649 non-null    int64 
 14  failures    649 non-null    int64 
 15  schoolsup   649 non-null    object
 16  famsup      649 non-null    object
 17  paid        649 non-null    object
 18  activities  649 non-null    object
 19  nursery     649 non-null    object
 20  higher    

In [6]:
Data = data1.append(data2, ignore_index=True)

In [7]:
Data

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0        GP   F   18       U     GT3       A     4     4   at_home   teacher   
1        GP   F   17       U     GT3       T     1     1   at_home     other   
2        GP   F   15       U     LE3       T     1     1   at_home     other   
3        GP   F   15       U     GT3       T     4     2    health  services   
4        GP   F   16       U     GT3       T     3     3     other     other   
...     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
1039     MS   M   20       U     LE3       A     2     2  services  services   
1040     MS   M   17       U     LE3       T     3     1  services  services   
1041     MS   M   21       R     GT3       T     1     1     other     other   
1042     MS   M   18       R     LE3       T     3     2  services     other   
1043     MS   M   19       U     LE3       T     1     1     other   at_home   

      ... famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0     ...      4        3      4     1     1      3        4   0  11  11  
1     ...      5        3      3     1     1      3        2   9  11  11  
2     ...      4        3      2     2     3      3        6  12  13  12  
3     ...      3        2      2     1     1      5        0  14  14  14  
4     ...      4        3      2     1     2      5        0  11  13  13  
...   ...    ...      ...    ...   ...   ...    ...      ...  ..  ..  ..  
1039  ...      5        5      4     4     5      4       11   9   9   9  
1040  ...      2        4      5     3     4      2        3  14  16  16  
1041  ...      5        5      3     3     3      3        3  10   8   7  
1042  ...      4        4      1     3     4      5        0  11  12  10  
1043  ...      3        2      3     3     3      5        5   8   9   9  

[1044 rows x 33 columns]

In [8]:
Data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1044 entries, 0 to 1043
Data columns (total 33 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   school      1044 non-null   object
 1   sex         1044 non-null   object
 2   age         1044 non-null   int64 
 3   address     1044 non-null   object
 4   famsize     1044 non-null   object
 5   Pstatus     1044 non-null   object
 6   Medu        1044 non-null   int64 
 7   Fedu        1044 non-null   int64 
 8   Mjob        1044 non-null   object
 9   Fjob        1044 non-null   object
 10  reason      1044 non-null   object
 11  guardian    1044 non-null   object
 12  traveltime  1044 non-null   int64 
 13  studytime   1044 non-null   int64 
 14  failures    1044 non-null   int64 
 15  schoolsup   1044 non-null   object
 16  famsup      1044 non-null   object
 17  paid        1044 non-null   object
 18  activities  1044 non-null   object
 19  nursery     1044 non-null   object
 20  higher  

# 1. Сколько в датасете объектов и признаков? Дать описание каждому признаку, если оно есть.

In [9]:
Data.columns


Index(['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu',
       'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime',
       'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery',
       'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc',
       'Walc', 'health', 'absences', 'G1', 'G2', 'G3'],
      dtype='object')

Объектов - 1044, Признаков - 33

Признаки:

1.school(школа) - бинарный признак: "GP" - Gabriel Pereira / "MS" - Mousinho da Silveira

2.sex(пол) - бинарный признак: "F" - female / "M" - male 

3.age(возраст) - числовой признак: 15<= age <= 22

4.address(тип населённого пункта) - бинарный признак:"U" - urban / "R" - rural 

5.famsize(количество человек в семье) - бинарный признак: "LE3" - <=3 / "GT3" - >3

6.Pstatus(состоят ли родители в браке) - бинарный признак: "T" - living together / "A" - Apart

7.Medu(образование матери) - категориальный признак: "0" - нет образования, "1" - начальное образование(закончила 4 класса),
"2" - неполное среднее(закончила 9 классов), "3" - среднее образование(закончила 11 классов),"4" - высшее образование

8.Fedu(образование отца) - категориальный признак: "0" - нет образования, "1" - начальное образование(закончил 4 класса),
"2" - неполное среднее(закончил 9 классов), "3" - среднее образование(закончил 11 классов),"4" - высшее образование

9.Mjob(работа матери) - категориальный признак: "teacher", "health" - care related, "services" - administrative or police, "at_home", "other"

10.Fjob(работа отца) - категориальный признак: "teacher", "health" - care related, "services" - administrative or police, "at_home", "other"

11.reason(причина выбора школы) - категориальный признак: "close to home", "school reputation", "course preference", "other"

12.guardian(опекун) - категориальный признак: "mother", "father", "other"

13.traveltime(сколько добираться до школы) - категориальный признак: "1" - < 15 минут, "2" - 15 минут <= traveltime <= 30 минут, "3" - 30 минут <= traveltime <= 1 час, "4" - > 1 час

14.studytime(сколько часов в неделю уделяется учёбе) - категориальный признак: "1" - < 2 часов, 
"2" - 2 часа <= studytime <= 5 часов, "3" - 5 часов <= traveltime <= 10 часов, "4" - > 10 часов

15.failures(сколько раз ученик оставался на второй год) - категориальный признак: "1" - 1 раз, "2" - 2 раза, "3" - 3 раза, 
"4" - > 3 раз

16.schoolsup(ученик имеет дополнительную образовательную поддержку) - бинарный признак: "yes" / "no"

17.famsup(ученик получает помощь в образовании со стороны семьи) - бинарный признак: "yes" / "no"

18.paid(ученик посещает дополнительные платные занятия по предмету курса "Math / Portuguese") - бинарный признак: "yes" / "no"

19.activities(участвует во внеклассной деятельности) - бинарный признак: "yes" / "no"

20.nursery(посещал детский сад) - бинарный признак: "yes" / "no"

21.higher(заинтересован в получении высшего образования) - бинарный признак: "yes" / "no"

22.internet(дома есть подключение к интернету) - бинарный признак: "yes" / "no"

23.romantic(состоит в романтических отношениях) -  бинарный признак: "yes" / "no"

24.famrel(взаимоотношения в семье) - категориальный признак: шкала от 1 до 5, где "1" - very bad, "5" - excellent

25.freetime(свободное время после школы) - категориальный признак: шкала от 1 до 5, где "1" - very low, "5" - very high

26.goout(прогулки на улице) - категориальный признак: шкала от 1 до 5, где "1" - very low, "5" - very high

27.Dalc(употребление алкоголя в будний день) - категориальный признак: шкала от 1 до 5, где "1" - very low, "5" - very high

28.Walc(употребление алкоголя в выходные дни) - категориальный признак: шкала от 1 до 5, где "1" - very low, "5" - very high

29.health(состояние здоровья) - категориальный признак: шкала от 1 до 5, где "1" - very bad, "5" - very good

30.absences(количество прогулов занятий в школе) -  числовой признак: 0 <= absences <= 93

31.G1(оценка за первый период) - числовой признак: 0 <= G1 <= 20

32.G2(оценка за второй период) - числовой признак: 0 <= G2 <= 20

33.G3(итоговая оценка) - числовой признак: 0 <= G3 <= 20

# 2. Сколько в Датасете категориальных признаков, какие именно?

В данном Датасете 15 категориальных признаков:

1.Medu(образование матери) 

2.Fedu(образование отца) 

3.Mjob(работа матери)

4.Fjob(работа отца) 

5.reason(причина выбора школы) 

6.guardian(опекун) 

7.traveltime(сколько добираться до школы) 

8.studytime(сколько часов в неделю уделяется учёбе) 

9.failures(сколько раз ученик оставался на второй год) 

10.famrel(взаимоотношения в семье)

11.freetime(свободное время после школы)

12.goout(прогулки на улице) 

13.Dalc(употребление алкоголя в будний день) 

14.Walc(употребление алкоголя в выходные дни) 

15.health(состояние здоровья) 

# 3. Столбец с макимальным количеством уникальных значений категориального признака?

In [10]:
Data_test = ['Medu', 'Fedu', 'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime', 'failures', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health']
for column in Data.columns:
    if column in Data_test:
        print(column, Data[column].nunique())

Medu 5
Fedu 5
Mjob 5
Fjob 5
reason 4
guardian 3
traveltime 4
studytime 4
failures 4
famrel 5
freetime 5
goout 5
Dalc 5
Walc 5
health 5


В данном Датасете 10 столбцов с максимальным количеством уникальных значений категориального признака:

1)Medu = 5;
2)Fedu = 5;
3)Mjob = 5;
4)Fjob = 5;
5)famrel = 5;
6)freetime = 5;
7)goout = 5;
8)Dalc = 5;
9)Walc = 5;
10)health = 5;

# 4. Есть ли бинарные признаки?

В данном Датасете 13 бинарных признаков:

1.school(школа) 

2.sex(пол)

3.address(тип населённого пункта)

4.famsize(количество человек в семье)

5.Pstatus(состоят ли родители в браке)

6.schoolsup(ученик имеет дополнительную образовательную поддержку)

7.famsup(ученик получает помощь в образовании со стороны семьи) 

8.paid(ученик посещает дополнительные платные занятия по предмету курса "Math / Portuguese")

9.activities(участвует во внеклассной деятельности)

10.nursery(посещал детский сад)

11.higher(заинтересован в получении высшего образования) 

12.internett(дома есть подключение к интернету) 

13.romantic(состоит в романтических отношениях)

# 5.Какие числовые признаки?

1.age(возраст)

2.absences(количество прогулов занятий в школе) 

3.G1(оценка за первый период) 

4.G2(оценка за второй период) 

5.G3(итоговая оценка) 

# 6. Есть ли пропуски?

In [11]:
Data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1044 entries, 0 to 1043
Data columns (total 33 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   school      1044 non-null   object
 1   sex         1044 non-null   object
 2   age         1044 non-null   int64 
 3   address     1044 non-null   object
 4   famsize     1044 non-null   object
 5   Pstatus     1044 non-null   object
 6   Medu        1044 non-null   int64 
 7   Fedu        1044 non-null   int64 
 8   Mjob        1044 non-null   object
 9   Fjob        1044 non-null   object
 10  reason      1044 non-null   object
 11  guardian    1044 non-null   object
 12  traveltime  1044 non-null   int64 
 13  studytime   1044 non-null   int64 
 14  failures    1044 non-null   int64 
 15  schoolsup   1044 non-null   object
 16  famsup      1044 non-null   object
 17  paid        1044 non-null   object
 18  activities  1044 non-null   object
 19  nursery     1044 non-null   object
 20  higher  

Основываясь на результате info(), мы видим, что пропусков в датасете нет (3-ий столбец)

# 7. Сколько объектов с пропусками?

Ответ: ни одного

# 8. Столбец с максимальным количеством пропусков?

Ответ: ни одного

# 9. Есть ли на ваш взгляд выбросы, аномальные значения?

In [12]:
for column in Data.columns:
    print(Data[column].value_counts())

GP    772
MS    272
Name: school, dtype: int64
F    591
M    453
Name: sex, dtype: int64
16    281
17    277
18    222
15    194
19     56
20      9
21      3
22      2
Name: age, dtype: int64
U    759
R    285
Name: address, dtype: int64
GT3    738
LE3    306
Name: famsize, dtype: int64
T    923
A    121
Name: Pstatus, dtype: int64
4    306
2    289
3    238
1    202
0      9
Name: Medu, dtype: int64
2    324
1    256
3    231
4    224
0      9
Name: Fedu, dtype: int64
other       399
services    239
at_home     194
teacher     130
health       82
Name: Mjob, dtype: int64
other       584
services    292
teacher      65
at_home      62
health       41
Name: Fjob, dtype: int64
course        430
home          258
reputation    248
other         108
Name: reason, dtype: int64
mother    728
father    243
other      73
Name: guardian, dtype: int64
1    623
2    320
3     77
4     24
Name: traveltime, dtype: int64
2    503
1    317
3    162
4     62
Name: studytime, dtype: int64
0    861
1  

Аномальных значений и выбросов не наблюдается

# 10. Столбец с максимальным средним значением после нормировки признаков через стандартное отклонение?

Нормируем признак для числовых признаков через стандартное отклонение:

In [13]:
 Data['age'] = (Data['age'] - Data['age'].mean())/(math.sqrt(Data['age'].var()))

In [14]:
 Data['absences'] = (Data['absences'] - Data['absences'].mean())/(math.sqrt(Data['absences'].var()))   

In [15]:
 Data['G1'] = (Data['G1'] - Data['G1'].mean())/(math.sqrt(Data['G1'].var())) 

In [16]:
 Data['G2'] = (Data['G2'] - Data['G2'].mean())/(math.sqrt(Data['G2'].var())) 

In [17]:
 Data['G3'] = (Data['G3'] - Data['G3'].mean())/(math.sqrt(Data['G3'].var())) 

Преобразуем текстовые данные в числовые:

In [18]:
array = ['school', 'sex', 'address', 'famsize', 'Pstatus', 'Mjob', 'Fjob', 'reason', 'guardian', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic']
for column in array:
    Set = set(Data[column])
    i = 0
    for item in Set:
        Data[column] = Data[column].replace(item, i)
        i = i + 1
    
    

In [19]:
Data

school  sex       age  address  famsize  Pstatus  Medu  Fedu  Mjob  \
0          1    0  1.027397        0        0        0     4     4     1   
1          1    0  0.220929        0        0        1     1     1     1   
2          1    0 -1.392007        0        1        1     1     1     1   
3          1    0 -1.392007        0        0        1     4     2     4   
4          1    0 -0.585539        0        0        1     3     3     2   
...      ...  ...       ...      ...      ...      ...   ...   ...   ...   
1039       0    1  2.640333        0        1        0     2     2     0   
1040       0    1  0.220929        0        1        1     3     1     0   
1041       0    1  3.446801        1        0        1     1     1     2   
1042       0    1  1.027397        1        1        1     3     2     0   
1043       0    1  1.833865        0        1        1     1     1     2   

      Fjob  ...  famrel  freetime  goout  Dalc  Walc  health  absences  \
0        3  ...       4         3      4     1     1       3 -0.070027   
1        2  ...       5         3      3     1     1       3 -0.392087   
2        2  ...       4         3      2     2     3       3  0.252034   
3        0  ...       3         2      2     1     1       5 -0.714147   
4        2  ...       4         3      2     1     2       5 -0.714147   
...    ...  ...     ...       ...    ...   ...   ...     ...       ...   
1039     0  ...       5         5      4     4     5       4  1.057185   
1040     0  ...       2         4      5     3     4       2 -0.231057   
1041     2  ...       5         5      3     3     3       3 -0.231057   
1042     2  ...       4         4      1     3     4       5 -0.714147   
1043     1  ...       3         2      3     3     3       5  0.091004   

            G1        G2        G3  
0    -3.758673 -0.074936 -0.088479  
1    -0.741974 -0.074936 -0.088479  
2     0.263592  0.533879  0.170267  
3     0.933969  0.838287  0.687758  
4    -0.071597  0.533879  0.429013  
...        ...       ...       ...  
1039 -0.741974 -0.683750 -0.605971  
1040  0.933969  1.447102  1.205250  
1041 -0.406786 -0.988158 -1.123463  
1042 -0.071597  0.229472 -0.347225  
1043 -1.077163 -0.683750 -0.605971  

[1044 rows x 33 columns]

Найдем максимальное среднее значение:

In [20]:
max_average = -math.inf
for column in array:
    if Data[column].mean() > max_average:
        max_average = Data[column].mean()
print(column,max_average)

romantic 1.6379310344827587


Ответ: столбец с максимальным средним значением после нормировки признаков - romantic

# 11. Столбец с целевым признаком?

Поскольку нам будет необходимо построить прогноз итогового бала учащегося, то целевой признак - G3

# 12. Сколько объектов попадает в тренировочную выборку при использовании train_test_split с параметрами test_size = 0.3, random_state = 42?

Выделим целевой признак(G3) и удалим его из Датасета

In [21]:
Data2 = Data.G3
train = Data.drop('G3', axis = 1)

Разделяем данные на обучающую и тестовую выборку

In [22]:
X_train, X_test, Y_train, Y_test = train_test_split(train,Data2, test_size = 0.3, random_state = 42)
train.shape

(1044, 32)

In [23]:
X_train

school  sex       age  address  famsize  Pstatus  Medu  Fedu  Mjob  Fjob  \
447       0    1  0.220929        1        0        1     2     1     2     2   
694       1    0 -1.392007        0        1        0     4     3     2     2   
448       0    0 -0.585539        1        0        1     4     4     3     3   
934       1    1  0.220929        0        0        1     1     1     2     2   
582       0    0  1.833865        0        1        0     1     1     1     2   
..      ...  ...       ...      ...      ...      ...   ...   ...   ...   ...   
87        1    0 -1.392007        0        0        1     4     2     2     2   
330       1    0  1.027397        0        0        1     2     1     2     2   
466       0    1 -1.392007        0        0        1     3     1     2     0   
121       1    1 -1.392007        0        0        1     2     2     0     0   
860       1    1  0.220929        0        1        1     4     4     0     2   

     ...  romantic  famrel  freetime  goout  Dalc  Walc  health  absences  \
447  ...         1       5         5      5     5     5       3  0.574094   
694  ...         1       5         2      2     1     1       5  0.574094   
448  ...         1       4         2      2     1     1       4  0.252034   
934  ...         0       4         3      3     1     2       4 -0.392087   
582  ...         0       1         4      4     1     1       5 -0.714147   
..   ...       ...     ...       ...    ...   ...   ...     ...       ...   
87   ...         0       5         3      3     1     3       1 -0.070027   
330  ...         1       4         3      1     1     1       5  0.896154   
466  ...         0       3         2      3     1     3       4 -0.714147   
121  ...         0       5         5      4     1     2       5  0.252034   
860  ...         1       5         3      5     4     5       3  1.379245   

           G1        G2  
447 -1.077163 -0.379343  
694 -1.077163 -0.988158  
448  1.604347  1.447102  
934  0.263592 -0.379343  
582 -1.747540 -0.988158  
..        ...       ...  
87   1.269158  1.142694  
330  0.263592  0.533879  
466 -0.406786 -0.683750  
121  0.933969  0.533879  
860  0.263592  0.229472  

[730 rows x 32 columns]

Ответ: В тренировочную выборку попадут 730 объектов

# 13. Между какими признаками наблюдается линейная зависимость (корреляция)?

Проверим коэффициент VIF для всех признаков:

In [24]:
vif = pandas.DataFrame()
vif["column_name"] = train.columns
vif["VIF"] = [variance_inflation_factor(train.values, i) for i in range(len(train.columns))]

Если коэффициент VIF > 10, то существует линейная зависимость:

In [25]:
vif.loc[vif['VIF'] > 10.0]

column_name        VIF
6         Medu  13.219419
7         Fedu  10.361689
20      higher  13.300180
23      famrel  17.835771
24    freetime  12.320756
25       goout  11.915192

Линейная зависимость наблюдается у следующий признаков: Medu, Fedu, higher, famrel, freetime, goout

# 14. Сколько признаков достаточно для объяснения 90% дисперсии после применения метода PCA?

Сначала стандартизируем данные:

In [26]:
New_Data = pandas.DataFrame(preprocessing.scale(X_train),columns = X_train.columns)

In [27]:
for i in range(1, len(New_Data.columns)):
    pca = PCA(n_components=i)
    pca.fit(New_Data)
    print(i, sum(pca.explained_variance_ratio_))

1 0.1140792856803792
2 0.1861273162471954
3 0.24625301865414223
4 0.2970196284652767
5 0.34274495923823073
6 0.38681289185701295
7 0.42991969421827436
8 0.46698036302832535
9 0.503066507316595
10 0.5369819286303772
11 0.5695758046027369
12 0.6005581272214687
13 0.6311432737817644
14 0.6593436005533172
15 0.6874027792939384
16 0.7148310655588254
17 0.741775434697729
18 0.7670136462971385
19 0.7908037967133287
20 0.8138810656842929
21 0.8352012281006531
22 0.8561053448966495
23 0.8759958227774755
24 0.8952576737823409
25 0.9134945809494734
26 0.9306227964626033
27 0.9469122636309413
28 0.962631472917743
29 0.9764994931830396
30 0.9865976511290891
31 0.9956564588013136


Для объяснения 90% дисперсии - достаточно 24 признаков

# 15. Какой признак вносит наибольший вклад в первую компоненту?

In [28]:
pca = PCA(n_components=24)
X = pca.fit(New_Data)

First_component = pca.components_[0]
i = 0
result = First_component[i]

for j in range(0, len(First_component)):
    if  abs(First_component[j]) >  abs(result):
        result = First_component[j]
        i = j
print(i, result)

31 -0.33494913434254237


Наибольший вклад в первую компоненту вносит 31-ый признак

In [29]:
New_Data.iloc[:, 31]

0     -0.390550
1     -1.004992
2      1.452777
3     -0.390550
4     -1.004992
         ...   
725    1.145556
726    0.531114
727   -0.697771
728    0.531114
729    0.223893
Name: G2, Length: 730, dtype: float64

Ответ: это признак G2